src/beatree/ops/reconstruction.rs

```diff
@@ -1,136 +1,95 @@
-    leaf,
-    Key,
 };
 
 // 256K
 const SEQ_READ_CHUNK_LEN: usize = BRANCH_NODE_SIZE * 64;
 type SeqReadChunk = [u8; SEQ_READ_CHUNK_LEN];
 
 /// Reconstruct the upper branch nodes of the btree from the bottom branch nodes and the leaf nodes.
-/// This places all branches into the BNP and returns an index into all BBNs. 
+/// This places all branches into the BNP and returns an index into all BBNs.
 pub fn reconstruct(
-    sync_seqn: u32,
-    bn_fd: File, 
+    bn_fd: File,
     bnp: &mut branch::BranchNodePool,
+    bbn_freelist: &BTreeSet<PageNumber>,
 ) -> Result<Index> {
-    let mut branch_seqns: HashMap<u64, (u32, Option<(Key, BranchId)>)> = HashMap::new();
     let mut index = Index::default();
-    let mut displaced = Vec::new();
 
     for seq_chunk in read_sequential(bn_fd)? {
         let seq_chunk = seq_chunk?;
         let nodes = seq_chunk.chunks(BRANCH_NODE_SIZE);
         for node in nodes {
             let view = branch::BranchNodeView::from_slice(node);
-            if view.sync_seqn() > sync_seqn { continue }
+            if bbn_freelist.contains(&view.bbn_pn().into()) {
+                continue;
+            }
 
             // handle empty.
             if view.n() == 0 && node == [0; BRANCH_NODE_SIZE] {
-                continue
+                continue;
             }
 
-            if let Some((existing_sync_seqn, separator)) = branch_seqns.get(&view.bbn_seqn()) {
-                if *existing_sync_seqn < view.sync_seqn() {
-                    if let Some((separator, expected_branch)) = *separator {
-                        // remove the previous branch, handling a corner case where it's already
-                        // been displaced by some other branch that shares the exact same
-                        // separator.
-                        // UNWRAP: indices always correspond to `Some` entries.
-                        let branch = index.remove(&separator).unwrap();
-                        if branch != expected_branch {
-                            index.insert(separator, branch);
-                        } else {
-                            bnp.release(branch);
-                        }
-                    }
-                } else {
-                    continue
-                }
+            let new_branch_id = bnp.allocate();
+
+            // UNWRAP: just allocated
+            bnp.checkout(new_branch_id)
+                .unwrap()
+                .as_mut_slice()
+                .copy_from_slice(node);
+
+            let mut separator = [0u8; 32];
+            {
+                let prefix = view.prefix();
+                let separator = separator.view_bits_mut::<Lsb0>();
+                separator[..prefix.len()].copy_from_bitslice(prefix);
+                let first = view.separator(0);
+                separator[prefix.len()..prefix.len() + first.len()].copy_from_bitslice(first);
+            }
+            if let Some(_displaced_branch_id) = index.insert(separator, new_branch_id) {
```

Comment:
the displacement logic is no longer necessary given that we have the freelist loaded up-front.

Reply:
should be updated in #44